backend/ipmi.pm

```diff
@@ -32,7 +32,7 @@ sub ipmitool ($self, $cmd, %args) {
     my @tries = (1 .. $args{tries});
     for (@tries) {
         $ret = IPC::Run::run(\@cmd, \$stdin, \$stdout, \$stderr);
-        if ($ret == 0) {
+        if ($ret eq 0) {
```

Comment:
> if $ret is empty or undefined, $ret == 0 is true, we need use "eq".

If `$ret` is `undef` the new code will produce a warning. Considering you say `undef` is to be expected this should probably be:

```suggestion
        if (defined $ret && $ret eq 0) {
```

It is also a bit weird to use `eq` here. Can `$ret` really be empty? If not I would use:

```suggestion
        if (defined $ret && $ret == 0) {
```

Reply:
Thanks @Martchus! I believe double equal can not work as expected.

Please refer to followed sample when $ret is empty string, the code will print "Hello"
```
my $ret = "";

if (defined $ret && $ret == 0) {
    print "Hello";                          #The code can print Hello -> incorrect
}
```

Let us verify `eq` as following
```
my $ret = 0;

if (defined $ret && $ret eq 0) {
    print 'Hello';                           # print Hello -> correct
}
-----------
my $ret = "0";

if (defined $ret && $ret eq 0) {
    print 'Hello';                           # print Hello -> correct
}
-----------
my $ret = "";

if (defined $ret && $ret eq 0) {
    print 'Hello';                          # Not print Hello -> correct
}
```

Reply:
According to https://metacpan.org/pod/IPC::Run#RETURN-VALUES there are just 2 distinctions to make. All sub processed existed with a 0 exit code  (then `$ret` would be a truthy value) or at least one sub processed existed with a non-0 exit code.

So likely this should just be:

```suggestion
        if (!$ret) {
```

There's no use in making a distinction between `0`, `''` and `undef` because all are falsy values and have the same meaning.

Reply:
During previous debugging, I observed the $ret is empty string, see the cut part from http://10.67.129.4/tests/62560
```
[2023-09-21T15:48:50.141416+08:00] [debug] [pid:959] IPMI ret: 1
[2023-09-21T15:48:54.210818+08:00] [debug] [pid:959] IPMI ret: 1
[2023-09-21T15:48:58.211160+08:00] [debug] [pid:959] IPMI: Chassis Power is off
[2023-09-21T15:48:58.280219+08:00] [debug] [pid:959] IPMI ret:
[2023-09-21T15:49:02.281003+08:00] [info] [pid:959] ::: backend::baseclass::die_handler: Backend process died, backend errors are reported below in the following lines:
ipmitool -I lanplus -H 10.67.135.18 -U root -P [masked] chassis power on: Set Chassis Power Control to Up/On failed: Command not supported in present state at /usr/lib/os-autoinst/backend/ipmi.pm line 49.
```

`if (!$ret)` still make `if` condition is incorrectly true when $ret is empty string, please try the followed codes. 
```
my $ret = '';

if (!$ret) {
    print "Hello";                     #The code can print Hello -> incorrect and not expected.
}
```
I believe `==` is used for numerical only, `eq` may more widely compatible for '0', 0, -1, 1 and ''.

Feel free to correct with if any is wrong.

Reply:
If ret was an empty string (or rather *some* falsy value logged like an empty string) this means the command returned with a non-zero exit code. This is the exact same case when 0 was returned. So I really don't get why you don't want to treat them the same. But maybe I'm also misreading the documentation of `IPC::Run` (which I don't find very good/explicit).

Reply:
For ipmitool, returning `empty` string or non-zero value/string means failure, `0` or 0 means successfully.

Reply:
Martchus is right. I also see checking for equality of (pseudo-)boolean values as bad style. The correct way is to treat it as boolean so `if ($ret)` and `if (!ret)` or `… unless $ret`, etc.

For a bit of context, this is who other uses of IPC::Run::run look like in os-autoinst+openQA:

```
okurz@linux-28d7:~/local/os-autoinst/os-autoinst 0 (fix/tidy) $ git grep -A 5 'IPC.*run'
OpenQA/Isotovideo/Utils.pm:        IPC::Run::run(\@{$debugging_tools{$tool}}, \$stdin, \$stdout, \$stderr);
OpenQA/Isotovideo/Utils.pm-    }
OpenQA/Isotovideo/Utils.pm-}
OpenQA/Isotovideo/Utils.pm-
OpenQA/Isotovideo/Utils.pm-1;
--
backend/amt.pm:    $ret = IPC::Run::run(\@cmd, \$stdin, \$stdout, \$stderr);
backend/amt.pm-    chomp $stdout;
backend/amt.pm-    chomp $stderr;
backend/amt.pm-
backend/amt.pm-    die join(' ', @cmd) . ": $stderr" unless ($ret);
backend/amt.pm-    bmwqemu::diag("AMT: $stdout");
--
backend/ipmi.pm:        $ret = IPC::Run::run(\@cmd, \$stdin, \$stdout, \$stderr);
backend/ipmi.pm-        if ($ret == 0) {
backend/ipmi.pm-            $self->dell_sleep;
backend/ipmi.pm-            last;
backend/ipmi.pm-        } else {
backend/ipmi.pm-            sleep 4;
okurz@linux-28d7:~/local/os-autoinst/os-autoinst 0 (fix/tidy) $ cd ../openQA
okurz@linux-28d7:~/local/os-autoinst/openQA 0 (feature/spew) $ git grep -A 5 'IPC.*run'
lib/OpenQA/Schema/ResultSet/Jobs.pm:        IPC::Run::run(\@cmd, \undef, \$stdout, \$stderr);
lib/OpenQA/Schema/ResultSet/Jobs.pm-        next if $stderr;
lib/OpenQA/Schema/ResultSet/Jobs.pm-        push(@results, map { $_ =~ s/\..*$//; basename $_} split(/\n/, $stdout));
lib/OpenQA/Schema/ResultSet/Jobs.pm-    }
lib/OpenQA/Schema/ResultSet/Jobs.pm-    return \@results;
lib/OpenQA/Schema/ResultSet/Jobs.pm-}
--
lib/OpenQA/Utils.pm:        my $ipc_run_succeeded = IPC::Run::run($cmd, \$stdin, \$stdout, \$stderr);
lib/OpenQA/Utils.pm-        my $return_code = $?;
lib/OpenQA/Utils.pm-        chomp $stderr;
lib/OpenQA/Utils.pm-        if ($ipc_run_succeeded) {
lib/OpenQA/Utils.pm-            OpenQA::Log->can("log_$stdout_level")->($stdout);
lib/OpenQA/Utils.pm-            OpenQA::Log->can("log_$stderr_level")->($stderr);
--
lib/OpenQA/WebAPI/Controller/API/V1/Search.pm:        IPC::Run::run(\@cmd, \undef, \$stdout, \$stderr);
lib/OpenQA/WebAPI/Controller/API/V1/Search.pm-        return $self->render(json => {error => "Grep failed: $stderr"}, status => 400) if $stderr;
lib/OpenQA/WebAPI/Controller/API/V1/Search.pm-
lib/OpenQA/WebAPI/Controller/API/V1/Search.pm-        my $basename = $distri->basename;
lib/OpenQA/WebAPI/Controller/API/V1/Search.pm-        my $last_filename = '';
lib/OpenQA/WebAPI/Controller/API/V1/Search.pm-        my @lines = split("\n", $stdout);
--
lib/OpenQA/WebAPI/Plugin/ObsRsync/Task.pm:    eval { IPC::Run::run(\@cmd, \$stdin, \$stdout, \$error); $exit_code = $?; };
lib/OpenQA/WebAPI/Plugin/ObsRsync/Task.pm-    my $error_from_exception = $@;
lib/OpenQA/WebAPI/Plugin/ObsRsync/Task.pm-
lib/OpenQA/WebAPI/Plugin/ObsRsync/Task.pm-    $helper->unlock($project);
lib/OpenQA/WebAPI/Plugin/ObsRsync/Task.pm-    return $job->finish(0) if (!$exit_code);
lib/OpenQA/WebAPI/Plugin/ObsRsync/Task.pm-
--
lib/OpenQA/WebAPI/Plugin/ObsRsync/Task.pm:        eval { IPC::Run::run(\@cmd, \$stdin, \$stdout, \$error); $exit_code = $?; };
lib/OpenQA/WebAPI/Plugin/ObsRsync/Task.pm-        my $err = $@;
lib/OpenQA/WebAPI/Plugin/ObsRsync/Task.pm-        return ($exit_code, $error // $err);
lib/OpenQA/WebAPI/Plugin/ObsRsync/Task.pm-    };
lib/OpenQA/WebAPI/Plugin/ObsRsync/Task.pm-
lib/OpenQA/WebAPI/Plugin/ObsRsync/Task.pm-    my ($exit_code, $error) = $helper->for_every_batch($alias, $sub);
```

so in any other place there is a proper check for just true/false evaluation of the returned value or just catching any exceptions and looking into the actual exit code from `$?`